posts/2023-07-24_rounding/rounding.qmd

```diff
@@ -0,0 +1,168 @@
+
+-   [CAMIS](https://psiaims.github.io/CAMIS/) (Comparing Analysis Method Implementations in Software): A cross-industry initiative to document discrepant results between software.
+    [Rounding](https://psiaims.github.io/CAMIS/Comp/r-sas_rounding.html) is one of the comparisons, and there are much more [on this page](https://psiaims.github.io/CAMIS/)!
+
+## Round half up in R
+
+The motivation for having a 'round half up' function is clear: it's a widely used rounding method, but there are no such options available in base R.
+
+There are multiple forums that have discussed this topic, and quite a few functions already available.
+But which one to choose?
+
+The first time I needed to round half up in R, I chose the function from a [PHUSE paper](https://www.lexjansen.com/phuse-us/2020/ct/CT05.pdf) and applied it to my study.
+It works fine for a while until I encountered the following precision issue when double programming in R for TLGs made in SAS.
+
+### Numerical precision issue
+
+Example of rounding half up for 2436.845, with 2 decimal places:
+
+```{r, message = FALSE}
+# a function that rounds half up
+# exact copy from: https://www.lexjansen.com/phuse-us/2020/ct/CT05.pdf
+ut_round <- function(x, n = 0) {
+  # x is the value to be rounded
+  # n is the precision of the rounding
+  scale <- 10^n
+  y <- trunc(x * scale + sign(x) * 0.5) / scale
+  # Return the rounded number
+  return(y)
+}
+# round half up for 2436.845, with 2 decimal places
+ut_round(2436.845, 2)
+```
+
+The expected result is 2436.85, but the output rounds it down.
+Thanks to the community effort, there are already discussions and resolution available in a [StackOverflow post](https://stackoverflow.com/questions/12688717/round-up-from-5#comment110611119_12688836) -
+
+> There are numerical precision issues, e.g., `round2(2436.845, 2)` returns `2436.84.` Changing `z + 0.5 to z + 0.5 + sqrt(.Machine$double.eps)` seems to work for me. -- Gregor Thomas Jun 24, 2020 at 2:16
+
+-   `.Machine$double.eps` is a built-in constant in R that represents the smallest positive floating-point number that can be represented on the system (reference: [Machine Characteristics](https://www.math.ucla.edu/~anderson/rw1001/library/base/html/zMachine.html))
+
+-   The expression `+ sqrt(.Machine$double.eps)` is used to add a very small value to mitigate floating-point precision issues.
+
+-   For more information about computational precision and floating-point, see the following links -
+
+    -   R: [Why doesn't R think these numbers are equal?](https://cran.r-project.org/doc/FAQ/R-FAQ.html#Why-doesn_0027t-R-think-these-numbers-are-equal_003f)
+    -   SAS: [Numerical Accuracy in SAS Software](https://documentation.sas.com/doc/en/lrcon/9.4/p0ji1unv6thm0dn1gp4t01a1u0g6.htm)
+
+After the fix:
+
+```{r, message = FALSE}
+# revised rounds half up
```

Comment:
personally, I don't think we should do this. this will change a lot of behaviors, like `ut_round1(2436.84499999999999, 2)` gives 2436.85 while `round(2436.84499999999999, 2)` gives 2436.84 (which is correct)

Reply:
@clarkliming Thanks for the review! your example is a similar case to the one mentioned above: `a <- 1.5 - 0.5*sqrt(.Machine$double.eps)`

The reason for this behavior is that `2436.845 - 2436.84499999999999` is less than `sqrt(.Machine$double.eps)`, the adjustment in the function is not big enough to round it up. When we encounter this situation, maybe we could handle it by the following options -

1. accept this value and explain: `2436.845` and `2436.84499999999999` are nearly equal by using `all.equal()` function from base R with default `tolerance` value (default `tolerance = sqrt(.Machine$double.eps) `
      - the output from SAS `round(2436.84499999999999, 0.01)` is also `2436.85`

2. choose a smaller value than `sqrt(.Machine$double.eps)` or remove it to adjust this function

How would you do "round half up" in R if we don't choose those functions (`ut_round1`, `janitor::round_half_up`, ...)?

Reply:
I assume this function will be used deeply inside other functions and large-scale used, so controlling the tolerance manually is not likely(which means that users can only apply one tolerance to all the numeric results obtained).

I will split your question into two: one is how do we deal with round half up, the other is how do we deal with rounding error.

My response to these two questions is the same: no action is needed.

For question 1, R is following IEEE 754, with default bankers rounding. What SAS provided seems reasonable, however, R is not wrong. This is just a different method. R does not need to copy SAS. 

For question 2, about the accuracy loss in computation, I will also do nothing. This is an issue by design, and not possible to perfectly solve at the moment. 

Both issues won't affect the interpretation of the result

So for me, introducing the fact that R and SAS are different in rounding should be sufficient.

Come back to the issue. The function `ut_round1` tries to solve the rounding error arising from the design, but it did not. So it is important to also mention this point. (also, this issue is not possible to solve in the current design, this fix only fixes one part of the issue)

Reply:
I'm not trying to to claim that base R's round function is wrong or we should to strictly mimic SAS's behavior, but to provide round half up options in R. Similar as the intention of providing [rounding option in Tplyr](https://atorus-research.github.io/Tplyr/articles/options.html#ibm-rounding), offering users more flexibility - there may be contexts where 'round half up' is required.

What do you think of using the following examples to highlight those round_half_up functions do not offer the same level of precision and accuracy as the base R round function? 
Please feel free to update this post directly, or let me know how to clarify this point :)
```r
> a <- 1.5 - 0.5*sqrt(.Machine$double.eps)
> b <- 1.5 - 0.5*.Machine$double.eps
> janitor::round_half_up(a)
[1] 2
> janitor::round_half_up(b)
[1] 2
> round(a)
[1] 1
# base round reaches the precision limit:
> round(b)
[1] 2
```